ttl: show table name for jobs in SHOW SCHEDULES

chrisseto opened the pull request:
Previously, `SHOW SCHEDULES` would only show a table's ID for row level TTL jobs. For the convenience of debugging, this commit upgrades the `label` column to contain the table's name instead of its ID.

Rather than performing table name resolution at query time, the job name is set at creation time and updated by the schema change during table renames. This minimizes the likelihood of `SHOW SCHEDULES` being rendered unusable when a cluster is in an unstable state.

Epic: CRDB-18322
Fixes: #93774

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/107294)
<!-- Reviewable:end -->

chrisseto commented:
I was a bit uncertain about how to best handle certain cases. These are marked as "TODO" and I intend to resolve them before merging. I would appreciate any input!

Xiang-Gu commented on pkg/sql/rename_table.go:
```diff
@@ -232,6 +233,29 @@ func (n *renameTableNode) startExec(params runParams) error {
+	if tableDesc.HasRowLevelTTL() {
+		const (
+			opname = `update ttl job names`
+			query  = `update system.scheduled_jobs set schedule_name = 'row-level-ttl-' || $1 WHERE schedule_id = $2`
```
nit: is it better to use captials and some parenthesis here like
```
UPDATE system.scheduled_jobs SET schedule_name = ('row-level-ttl-' || $1) WHERE schedule_id = $2
```

chrisseto replied:
Done.

dt commented on pkg/sql/create_table.go:
```diff
@@ -2426,27 +2425,26 @@ func newTableDesc(
 	})
 
-	if err := sj.SetSchedule(ttl.DeletionCronOrDefault()); err != nil {
+	// TODO: Is it better to check that RowLevelTTL is not nil and return an error?
```
nit: `TODO(username): `

chengxiong-ruan replied:
+1 to dt's point

Also, If we're setting schedule for a nil TTL, it's a critical bug that we should let it fail hard. but it's not nice to panic. We could return a assertion error with `errors.AssertionFailedf` to feel safe.

chrisseto replied:
Sorry for the confusion. This was [intended to be a discussion at review time](https://github.com/cockroachdb/cockroach/pull/107294#issuecomment-1644436745). I'll use `TODO(BEFORE MERGING)` in the future!

> Also, If we're setting schedule for a nil TTL, it's a critical bug that we should let it fail hard. but it's not nice to panic. We could return a assertion error with errors.AssertionFailedf to feel safe.

Agreed that panicking isn't great. More generally, is the style in CRDB is to assume that programmers are doing the right thing or is it to be paranoid? Because of how this function got refactored and how go's type system works, callers could accidentally cause nil dereferences. I find myself running into these types of cases frequently. In general I prefer to write types in such a way that this can't happen but that's not always on option (See: this PR). My personal rule is to be paranoid at boundaries and expect well shaped data inside of said boundaries. I don't really have a sense of what is or is not considered to be a boundary yet. `sql.CreateRowLevelTTLScheduledJob` passes `tblDesc` right through, so we could do a check at the package level but this metaphorical boundary might exist somewhere else instead. What would either of you do for this specific situation and does that follow a more generalizable rule?

chengxiong-ruan replied:
mhm, I think you already made good points here, we can do the check in `CreateRowLevelTTLScheduledJob` given it's a public function at the package level. I personally often makes assumptions in non-public functions, of course sometimes do such kind of validations in a helper.

chrisseto replied:
Thanks! Added the check to `CreateRowLevelTTLScheduleJob`.

chengxiong-ruan commented on pkg/sql/logictest/testdata/logic_test/row_level_ttl:
```diff
@@ -351,12 +351,44 @@ SELECT crdb_internal.validate_ttl_scheduled_jobs()
 
 subtest end
 
+# Ensure that SHOW SCHEDULES's label column is updated to reflect the new table
```
Could you add a test case where the schedule name uses id (mock with a manual update in test) and it's changed to use name after renaming the table? Just want to make sure your update query works.

chrisseto replied:
Done.

chengxiong-ruan commented on pkg/sql/rename_table.go:
```diff
@@ -232,6 +233,29 @@ func (n *renameTableNode) startExec(params runParams) error {
 
+	// If this table has row level ttl enabled, update the schedule_name of all
+	// row-level-ttl jobs with the name table name.
+	if tableDesc.HasRowLevelTTL() {
```
nit: maybe extract this logic into a separate helper function. I'm ok with what it's like now though.

chrisseto replied:
I don't think there's a super clear place to put it right now. Could we save the refactor until rename table is implemented in the declarative schema changer? I think having two consumers will make the eventual "home" of this logic much more clear.

chengxiong-ruan commented on pkg/sql/rename_table.go:
```diff
@@ -232,6 +233,29 @@ func (n *renameTableNode) startExec(params runParams) error {
+			opname = `update ttl job names`
+			query  = `update system.scheduled_jobs set schedule_name = 'row-level-ttl-' || $1 WHERE schedule_id = $2`
+		)
+		// TODO Should the number of modified rows be inspected here? Could log a
```
I don't think renaming a table should handle this. It's part of what the TTL infra itself should check. cc @ecwall to see if there is a strong guarantee here.

chrisseto replied:
Okay! I've removed this comment. Happy to revisit if Evan has opinions.

chengxiong-ruan reviewed (approved):
lgtm! only minor things

fqazi reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/107294)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @chengxiong-ruan, @chrisseto, @dt, @ecwall, and @rhu713)

___
*[`pkg/sql/rename_table.go` line 238 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/107294#gh-1270048805:-N_u7OPXEQAT67odfRay:b3ndz4a) ([raw file](https://github.com/cockroachdb/cockroach/blob/6d89dae40815d56f568055fa24cf9cb8440fce1c/pkg/sql/rename_table.go#L238)):*
<details><summary><i>Previously, chrisseto (Chris Seto) wrote…</i></summary><blockquote>

I don't think there's a super clear place to put it right now. Could we save the refactor until rename table is implemented in the declarative schema changer? I think having two consumers will make the eventual "home" of this logic much more clear.
</blockquote></details>

Lets move this over to MetaDataUpdater then the declarative side can easily consume it


<!-- Sent from Reviewable.io -->

fqazi reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/107294)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @chengxiong-ruan, @chrisseto, @dt, @ecwall, and @rhu713)

___
*[`pkg/sql/rename_table.go` line 238 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/107294#gh-1270048805:-N_u845f58K86wRXU2MX:bhl37yt) ([raw file](https://github.com/cockroachdb/cockroach/blob/6d89dae40815d56f568055fa24cf9cb8440fce1c/pkg/sql/rename_table.go#L238)):*
<details><summary><i>Previously, fqazi (Faizan Qazi) wrote…</i></summary><blockquote>

Lets move this over to MetaDataUpdater then the declarative side can easily consume it
</blockquote></details>

*MetadataUpdater


<!-- Sent from Reviewable.io -->

chengxiong-ruan commented:
*[`pkg/sql/rename_table.go` line 238 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/107294#gh-1270048805:-N_uSZm_EOE3U1ltvXYn:brfi1to) ([raw file](https://github.com/cockroachdb/cockroach/blob/6d89dae40815d56f568055fa24cf9cb8440fce1c/pkg/sql/rename_table.go#L238)):*
<details><summary><i>Previously, fqazi (Faizan Qazi) wrote…</i></summary><blockquote>

*MetadataUpdater
</blockquote></details>

Good idea. I like that.


<!-- Sent from Reviewable.io -->

chrisseto reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/107294)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @chengxiong-ruan, @dt, @ecwall, @fqazi, and @rhu713)

___
*[`pkg/sql/rename_table.go` line 238 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/107294#gh-1270048805:-Na7wnlT2xN5aemKpIFV:bsi74v4) ([raw file](https://github.com/cockroachdb/cockroach/blob/6d89dae40815d56f568055fa24cf9cb8440fce1c/pkg/sql/rename_table.go#L238)):*
<details><summary><i>Previously, chengxiong-ruan (Chengxiong Ruan) wrote…</i></summary><blockquote>

Good idea. I like that.
</blockquote></details>

Oh nice! That's the perfect spot. Moved.
___
*[`pkg/sql/rename_table.go` line 243 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/107294#gh-1270056218:-Na7wrs30h8H0S25gBux:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/6d89dae40815d56f568055fa24cf9cb8440fce1c/pkg/sql/rename_table.go#L243)):*
<details><summary><i>Previously, chrisseto (Chris Seto) wrote…</i></summary><blockquote>

Okay! I've removed this comment. Happy to revisit if Evan has opinions.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

chengxiong-ruan reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/107294#-:-Na8FS-q71VQeI6hNP8g:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @dt, @ecwall, @fqazi, and @rhu713)



<!-- Sent from Reviewable.io -->

rafiss reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/107294)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @chengxiong-ruan, @chrisseto, @dt, @ecwall, @fqazi, and @rhu713)

___
*[`pkg/sql/create_table.go` line 2434 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/107294#-NaCoFxG1V5mLw68YTkp:-NaCoFxG1V5mLw68YTkq:bnwatpk) ([raw file](https://github.com/cockroachdb/cockroach/blob/608e989209e797b4eadc8099331805f92bd063ca/pkg/sql/create_table.go#L2434)):*
> ```Go
> ) (*jobs.ScheduledJob, error) {
> 	sj := jobs.NewScheduledJob(env)
> 	sj.SetScheduleLabel(fmt.Sprintf("row-level-ttl-%s", tblDesc.GetName()))
> ```

nit: i think we may actually want both the name and ID.

the primary reason is that table names are not unique - the fully qualified name of `database_name.schema_name.table_name` is, but it doesn't seem like we have those other names easily available. (and it just seems like a headache to make all those RENAME operations get reflected in this schedule label too.)


<!-- Sent from Reviewable.io -->

chrisseto reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/107294)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @chengxiong-ruan, @dt, @ecwall, @fqazi, @rafiss, and @rhu713)

___
*[`pkg/sql/create_table.go` line 2434 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/107294#-NaCoFxG1V5mLw68YTkp:-NaDRGH73DTJquJKLdVy:b-myr6cr) ([raw file](https://github.com/cockroachdb/cockroach/blob/608e989209e797b4eadc8099331805f92bd063ca/pkg/sql/create_table.go#L2434)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

nit: i think we may actually want both the name and ID.

the primary reason is that table names are not unique - the fully qualified name of `database_name.schema_name.table_name` is, but it doesn't seem like we have those other names easily available. (and it just seems like a headache to make all those RENAME operations get reflected in this schedule label too.)
</blockquote></details>

Is there a need to make the label unique? The table ID is available in the "args" column, if disambiguation is needed. I do think that fully qualified name would be much more helpful, I just didn't know how to get it. :sweat_smile: 

```
root@localhost:36257/defaultdb2> SHOW SCHEDULES;

          id         |        label         | schedule_status |        next_run        |   state   | recurrence | jobsrunning | owner |            created            | on_wait |  on_error   |     command
---------------------+----------------------+-----------------+------------------------+-----------+------------+-------------+-------+-------------------------------+---------+-------------+-------------------
  883280186681032705 | sql-schema-telemetry | ACTIVE          | 2023-07-24 09:37:00+00 | pending   | 37 9 * * 1 |           0 | node  | 2023-07-17 20:35:20.996534+00 | SKIP    | RETRY_SCHED | {}
  883280186749747201 | sql-stats-compaction | ACTIVE          | 2023-07-21 21:00:00+00 | succeeded | @hourly    |           0 | node  | 2023-07-17 20:35:21.038388+00 | SKIP    | RETRY_SCHED | {}
  883280439706157057 | row-level-ttl-104    | ACTIVE          | 2023-07-21 21:00:00+00 | NULL      | @hourly    |           0 | root  | 2023-07-17 20:36:38.237751+00 | WAIT    | RETRY_SCHED | {"tableId": 104}
  885528023598104577 | row-level-ttl-tbl    | ACTIVE          | 2023-07-25 20:00:00+00 | NULL      | @hourly    |           0 | root  | 2023-07-25 19:08:26.403993+00 | WAIT    | RETRY_SCHED | {"tableId": 107}
  885528044534595585 | row-level-ttl-tbl    | ACTIVE          | 2023-07-25 20:00:00+00 | NULL      | @hourly    |           0 | root  | 2023-07-25 19:08:32.797117+00 | WAIT    | RETRY_SCHED | {"tableId": 108}
```

Hmmmmm, you know, I added this example to showcase my point but I think it's made me agree with you. The fully qualified name would be much nicer but we'd have to hook into database and schema renames as well and it would make the labels MUCH longer. Is`row-level-ttl-107-tbl` a good middle ground? If there's no limitation on the format of label, `row-level-ttl: 107 (tbl)` or something up that alley might be a bit more pleasant to read.


<!-- Sent from Reviewable.io -->

rafiss reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/107294)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @chengxiong-ruan, @chrisseto, @dt, @ecwall, @fqazi, and @rhu713)

___
*[`pkg/sql/create_table.go` line 2434 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/107294#-NaCoFxG1V5mLw68YTkp:-NaDTpH-0kXPS1R3nHtC:bxlmxiq) ([raw file](https://github.com/cockroachdb/cockroach/blob/608e989209e797b4eadc8099331805f92bd063ca/pkg/sql/create_table.go#L2434)):*
<details><summary><i>Previously, chrisseto (Chris Seto) wrote…</i></summary><blockquote>

Is there a need to make the label unique? The table ID is available in the "args" column, if disambiguation is needed. I do think that fully qualified name would be much more helpful, I just didn't know how to get it. :sweat_smile: 

```
root@localhost:36257/defaultdb2> SHOW SCHEDULES;

          id         |        label         | schedule_status |        next_run        |   state   | recurrence | jobsrunning | owner |            created            | on_wait |  on_error   |     command
---------------------+----------------------+-----------------+------------------------+-----------+------------+-------------+-------+-------------------------------+---------+-------------+-------------------
  883280186681032705 | sql-schema-telemetry | ACTIVE          | 2023-07-24 09:37:00+00 | pending   | 37 9 * * 1 |           0 | node  | 2023-07-17 20:35:20.996534+00 | SKIP    | RETRY_SCHED | {}
  883280186749747201 | sql-stats-compaction | ACTIVE          | 2023-07-21 21:00:00+00 | succeeded | @hourly    |           0 | node  | 2023-07-17 20:35:21.038388+00 | SKIP    | RETRY_SCHED | {}
  883280439706157057 | row-level-ttl-104    | ACTIVE          | 2023-07-21 21:00:00+00 | NULL      | @hourly    |           0 | root  | 2023-07-17 20:36:38.237751+00 | WAIT    | RETRY_SCHED | {"tableId": 104}
  885528023598104577 | row-level-ttl-tbl    | ACTIVE          | 2023-07-25 20:00:00+00 | NULL      | @hourly    |           0 | root  | 2023-07-25 19:08:26.403993+00 | WAIT    | RETRY_SCHED | {"tableId": 107}
  885528044534595585 | row-level-ttl-tbl    | ACTIVE          | 2023-07-25 20:00:00+00 | NULL      | @hourly    |           0 | root  | 2023-07-25 19:08:32.797117+00 | WAIT    | RETRY_SCHED | {"tableId": 108}
```

Hmmmmm, you know, I added this example to showcase my point but I think it's made me agree with you. The fully qualified name would be much nicer but we'd have to hook into database and schema renames as well and it would make the labels MUCH longer. Is`row-level-ttl-107-tbl` a good middle ground? If there's no limitation on the format of label, `row-level-ttl: 107 (tbl)` or something up that alley might be a bit more pleasant to read.
</blockquote></details>

right, your example is what i had in mind. i think both ID and table name is good to be both helpful enough and provide full disambiguation for the users who do have the same table names in different places. a convention we use sometimes is to put the ID in square brackets. something like `row-level-ttl: table_name [107]`


<!-- Sent from Reviewable.io -->

chrisseto reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/107294)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @chengxiong-ruan, @dt, @ecwall, @fqazi, @rafiss, and @rhu713)

___
*[`pkg/sql/create_table.go` line 2434 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/107294#-NaCoFxG1V5mLw68YTkp:-NaDdSXG3h1V9jVd-s2K:bhc3obi) ([raw file](https://github.com/cockroachdb/cockroach/blob/608e989209e797b4eadc8099331805f92bd063ca/pkg/sql/create_table.go#L2434)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

right, your example is what i had in mind. i think both ID and table name is good to be both helpful enough and provide full disambiguation for the users who do have the same table names in different places. a convention we use sometimes is to put the ID in square brackets. something like `row-level-ttl: table_name [107]`
</blockquote></details>

I like that! I'll go ahead and pull this all into a function in the TTL package so it's easier to refactor in the future.


<!-- Sent from Reviewable.io -->

chrisseto committed:
ttl: show table name for jobs in SHOW SCHEDULES

Previously, `SHOW SCHEDULES` would only show a table's ID for row level
TTL jobs. For the convenience of debugging, this commit upgrades the
`label` column to contain the table's name instead of its ID.

Rather than performing table name resolution at query time, the job name
is set at creation time and updated by the schema change during table
renames. This minimizes the likelihood of `SHOW SCHEDULES` being
rendered unusable when a cluster is in an unstable state.

Epic: CRDB-18322
Fixes: 93774

chrisseto commented:
Friendly ping @rafiss :)

rafiss reviewed (approved):
this looks good! thanks for the ping

> Reviewed 2 of 7 files at r1, 1 of 4 files at r2, 16 of 16 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/107294#-:-Nalap1KCpm-6P3B0pYg:bdw6ilm)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @chengxiong-ruan, @dt, @ecwall, @fqazi, and @rhu713)



<!-- Sent from Reviewable.io -->

chrisseto commented:
TFTR!

bors r+

chrisseto commented:
Hmm, I just realized that there likely isn't going to be a test that will fail once rename table is implemented in the declarative schema changer. I'll find or find an issue to make sure this doesn't get forgotten.
